src/openms/include/OpenMS/ANALYSIS/ID/IDBoostGraph.h

```diff
@@ -0,0 +1,540 @@
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Julianus Pfeuffer $
+// $Authors: Julianus Pfeuffer $
+// --------------------------------------------------------------------------
+
+#ifndef OPENMS_ANALYSIS_ID_IDBOOSTGRAPH_H
+#define OPENMS_ANALYSIS_ID_IDBOOSTGRAPH_H
+
+// define to get timings for connected components
+//#define INFERENCE_BENCH
+
+#include <OpenMS/ANALYSIS/ID/MessagePasserFactory.h> //included in BPI
+#include <OpenMS/CONCEPT/LogStream.h>
+#include <OpenMS/CONCEPT/Types.h>
+#include <OpenMS/KERNEL/StandardTypes.h>
+#include <OpenMS/METADATA/ExperimentalDesign.h>
+#include <OpenMS/METADATA/ProteinIdentification.h>
+#include <OpenMS/METADATA/PeptideIdentification.h>
+
+#include <vector>
+#include <unordered_map>
+#include <queue>
+
+#include <boost/function.hpp>
+#include <boost/graph/adjacency_list.hpp>
+#include <boost/graph/depth_first_search.hpp>
+#include <boost/graph/filtered_graph.hpp>
+#include <boost/graph/properties.hpp>
+#include <boost/variant.hpp>
+#include <boost/variant/detail/hash_variant.hpp>
+#include <boost/variant/static_visitor.hpp>
+
+namespace OpenMS
+{
+
+  /**
+   @brief Creates and maintains a boost graph based on the OpenMS ID datastructures
+
+   For finding connected components and applying functions to them.
+   Currently assumes that all PeptideIdentifications are from the ProteinID run that is given.
+   Please make sure this is right.
+   VERY IMPORTANT NOTE: If you add Visitors here, make sure they do not touch members of the
+   underlying ID objects that are responsible for the graph structure. E.g. the (protein/peptide)_hits vectors
+   or the lists in ProteinGroups. You can set information like scores or metavalues, though.
+
+   @ingroup Analysis_ID
+   */
+   //TODO Add OPENMS_DLLAPI everywhere
+  class OPENMS_DLLAPI IDBoostGraph
```

Comment:
put into `OpenMS::Internal` ?

Reply:
Also add a brief overview of the layers of the graph?